File: patterns/_tests/BehavioralPatterns/ChainOfResponsibility/ruby/integration.test.rb
# ============================================================
#  Integration Tests - Chain of Responsibility: Grammar Token Validator
#
#  What we test here:
#    Full token sequences accepted or rejected end-to-end
#    Invalid sequences fail at the correct chain link
#    Every token type correctly classified in a realistic equation
#    Chain extension does not break existing behaviour
#    Paren depth and decimal flags interact correctly
#
#  Pipeline tests:
#    CoR chain + TreeBuilder + Interpreter evaluate end-to-end
#    Operator precedence is respected through tree shape
#    Grouping with parentheses overrides precedence
#    Unary prefix functions nest correctly
#    Invalid token sequences raise RuntimeError
#
#  Run: ruby integration.test.rb
# ============================================================

require 'minitest/autorun'
require_relative '../../../../BehavioralPatterns/ChainOfResponsibility/ruby/ruby'

# ─────────────────────────────────────────────────────────────
#  Helper: feed a token sequence through the chain, updating
#  context automatically after each accepted token.
# ─────────────────────────────────────────────────────────────

def feed(chain, tokens)
  results = []
  ctx = ParseContext.new

  tokens.each do |tok|
    result = chain.handle(tok, ctx)
    results << result
    break if result.nil?

    token_type  = result[0]
    paren_depth = ctx.paren_depth
    has_decimal = ctx.has_decimal

    if tok == '('
      paren_depth += 1
      has_decimal  = false
    elsif tok == ')'
      paren_depth -= 1
    elsif %w[Exp Term Power].include?(token_type)
      has_decimal = false
    elsif tok == '.'
      has_decimal = true
    end

    ctx = ParseContext.new(trailing: tok, paren_depth: paren_depth, has_decimal: has_decimal)
  end

  results
end


class TestValidSequences < Minitest::Test
  def test_simple_addition
    r = feed(build_chain, %w[5 + 3 =])
    assert_equal %w[Digit Exp Digit Solve], r.map { |x| x[0] }
  end

  def test_order_of_operations
    r = feed(build_chain, %w[2 + 3 * 4 =])
    assert_equal %w[Digit Exp Digit Term Digit Solve], r.map { |x| x[0] }
  end

  def test_parenthesised
    r = feed(build_chain, ['(', '6', '-', '2', ')', '*', '3', '='])
    assert_equal %w[Factor Digit Exp Digit Factor Term Digit Solve], r.map { |x| x[0] }
  end

  def test_power_expression
    r = feed(build_chain, ['2', '**', '8', '='])
    assert_equal %w[Digit Power Digit Solve], r.map { |x| x[0] }
  end

  def test_function_wrapping
    r = feed(build_chain, %w[abs 9 =])
    assert_equal %w[Function Digit Solve], r.map { |x| x[0] }
  end

  def test_nested_functions
    r = feed(build_chain, %w[neg abs 7 =])
    assert_equal %w[Function Function Digit Solve], r.map { |x| x[0] }
  end

  def test_decimal_number
    r = feed(build_chain, ['3', '.', '1', '4', '='])
    assert_equal %w[Digit SoloDec Digit Digit Solve], r.map { |x| x[0] }
  end

  def test_floor_divide
    r = feed(build_chain, ['1', '7', '//', '5', '='])
    assert_equal %w[Digit Digit Term Digit Solve], r.map { |x| x[0] }
  end

  def test_modulo
    r = feed(build_chain, ['1', '7', '%', '5', '='])
    assert_equal %w[Digit Digit Term Digit Solve], r.map { |x| x[0] }
  end

  def test_complex
    r = feed(build_chain, %w[2 3 4 + ( 1 2 / 3 ) =])
    assert_equal %w[Digit Digit Digit Exp Factor Digit Digit Term Digit Factor Solve],
                 r.map { |x| x[0] }
  end

  def test_solve_only
    assert_equal ['Solve', '='], build_chain.handle('=', ParseContext.new)
  end
end


class TestInvalidSequences < Minitest::Test
  def test_leading_plus_rejected
    assert_nil feed(build_chain, ['+', '5'])[0]
  end

  def test_double_operator_rejected
    r = feed(build_chain, %w[1 2 + +])
    refute_nil r[0]
    assert_nil r[3]
  end

  def test_operator_after_open_paren
    assert_nil feed(build_chain, ['(', '+'])[1]
  end

  def test_unmatched_close_paren
    assert_nil build_chain.handle(')', ParseContext.new(paren_depth: 0))
  end

  def test_second_decimal_rejected
    assert_nil feed(build_chain, ['1', '.', '5', '.'])[3]
  end

  def test_digit_after_close_paren
    assert_nil feed(build_chain, ['(', '3', ')', '5'])[3]
  end

  def test_unknown_token
    assert_nil build_chain.handle('xyz', ParseContext.new(trailing: '5'))
  end

  def test_leading_multiply
    assert_nil build_chain.handle('*', ParseContext.new)
  end
end


class TestProcessedTokenValues < Minitest::Test
  def test_exp_adds_spaces
    assert_equal ' + ', build_chain.handle('+', ParseContext.new(trailing: '5'))[1]
  end

  def test_term_adds_spaces
    assert_equal ' * ', build_chain.handle('*', ParseContext.new(trailing: '3'))[1]
  end

  def test_power_adds_spaces
    assert_equal ' ** ', build_chain.handle('**', ParseContext.new(trailing: '2'))[1]
  end

  def test_factor_no_spaces
    assert_equal '(', build_chain.handle('(', ParseContext.new)[1]
  end

  def test_solo_decimal
    assert_equal '0.', build_chain.handle('.', ParseContext.new)[1]
  end

  def test_digit_as_is
    assert_equal '7', build_chain.handle('7', ParseContext.new)[1]
  end
end


class TestChainExtension < Minitest::Test
  def test_new_function_accepted
    ext = build_chain(functions: DEFAULT_FUNCTIONS + ['cbrt'])
    assert_equal ['Function', 'cbrt'], ext.handle('cbrt', ParseContext.new(trailing: '8'))
  end

  def test_new_function_rejected_in_default
    assert_nil build_chain.handle('cbrt', ParseContext.new(trailing: '8'))
  end

  def test_existing_functions_still_work
    ext = build_chain(functions: DEFAULT_FUNCTIONS + ['cbrt'])
    DEFAULT_FUNCTIONS.each do |fn|
      assert_equal ['Function', fn], ext.handle(fn, ParseContext.new(trailing: '5'))
    end
  end
end


class TestPipeline < Minitest::Test
  def setup
    @ctx = Context.new
  end

  def ev(tokens)
    parse_and_build(tokens).interpret(@ctx)
  end

  def test_simple_addition     = assert_equal 7,   ev(%w[3 + 4 =])
  def test_simple_subtraction  = assert_equal 5,   ev(%w[9 - 4 =])
  def test_simple_multiply     = assert_equal 42,  ev(%w[6 * 7 =])
  def test_simple_divide       = assert_equal 2.0, ev(%w[8 / 4 =])

  def test_precedence
    # 2 + 3 * 4 = 14
    assert_equal 14, ev(%w[2 + 3 * 4 =])
  end

  def test_parens_override
    assert_equal 20, ev(['(', '2', '+', '3', ')', '*', '4', '='])
  end

  def test_power       = assert_equal 256.0, ev(['2', '**', '8', '='])
  def test_floor_divide = assert_equal 3.0,  ev(['1', '7', '//', '5', '='])
  def test_modulo      = assert_equal 2.0,   ev(['1', '7', '%', '5', '='])
  def test_sqrt        = assert_equal 3.0,   ev(['sqrt', '9', '='])

  def test_abs_neg
    assert_equal 5.0, ev(['abs', 'neg', '5', '='])
  end

  def test_nested_unary
    assert_equal 7.0, ev(['abs', 'neg', 'abs', 'neg', '7', '='])
  end

  def test_pythagorean
    assert_equal 5.0, ev(['sqrt','(','3','**','2','+','4','**','2',')','='])
  end

  def test_multidigit
    assert_equal 234, ev(%w[2 3 4 + 0 =])
  end

  def test_decimal
    r = ev(['3', '.', '1', '4', '+', '0', '='])
    assert_in_delta 3.14, r, 1e-9
  end

  def test_invalid_token_raises
    assert_raises(RuntimeError) { parse_and_build(['+', '5']) }
  end

  def test_tree_str_reflects_structure
    tree = parse_and_build(['(', '2', '+', '3', ')', '*', '4', '='])
    assert_includes tree.to_s, '*'
    assert_includes tree.to_s, '+'
  end
end

File: patterns/_tests/BehavioralPatterns/ChainOfResponsibility/python/integration.test.py
# ============================================================
#  Integration Tests - Chain of Responsibility: Grammar Token Validator
#
#  What we test here:
#    · Full token sequences accepted or rejected end-to-end
#    · Invalid sequences fail at the correct chain link
#    · Every token type correctly classified in a realistic equation
#    · Chain extension doesn't break existing behaviour
#    · Paren depth and decimal flags interact correctly
#
#  Pipeline tests (TestPipeline):
#    · CoR chain + TreeBuilder + Interpreter evaluate end-to-end
#    · Operator precedence is respected through tree shape
#    · Grouping with parentheses overrides precedence
#    · Unary prefix functions nest correctly
#    · Invalid token sequences raise ValueError before evaluation
#
#  Run: pytest integration.test.py -v --import-mode=importlib
# ============================================================

import importlib.util, os, pytest

# Load CoR module by explicit path to avoid name collision with Interpreter's python.py
_cor_file = os.path.normpath(os.path.join(os.path.dirname(__file__),
    '../../../../BehavioralPatterns/ChainOfResponsibility/python/python.py'))
_spec = importlib.util.spec_from_file_location("cor", _cor_file)
_cor  = importlib.util.module_from_spec(_spec)
_spec.loader.exec_module(_cor)

ParseContext    = _cor.ParseContext
SolveHandler    = _cor.SolveHandler
ExpHandler      = _cor.ExpHandler
TermHandler     = _cor.TermHandler
PowerHandler    = _cor.PowerHandler
FactorHandler   = _cor.FactorHandler
FunctionHandler = _cor.FunctionHandler
DigitHandler    = _cor.DigitHandler
DEFAULT_FUNCTIONS = _cor.DEFAULT_FUNCTIONS
build_chain     = _cor.build_chain
TreeBuilder     = _cor.TreeBuilder
parse_and_build = _cor.parse_and_build

# Load Interpreter module by explicit path
_interp_file = os.path.normpath(os.path.join(os.path.dirname(__file__),
    '../../../../BehavioralPatterns/Interpreter/python/python.py'))
_spec2  = importlib.util.spec_from_file_location("interp", _interp_file)
_interp = importlib.util.module_from_spec(_spec2)
_spec2.loader.exec_module(_interp)


# ─────────────────────────────────────────────────────────────
#  Helper: feed a token sequence through the chain, updating
#  context automatically after each accepted token.
# ─────────────────────────────────────────────────────────────

def feed(chain, tokens: list[str]) -> list[tuple[str, str] | None]:
    results = []
    ctx = ParseContext()

    for tok in tokens:
        result = chain.handle(tok, ctx)
        results.append(result)

        if result is None:
            break

        token_type, _ = result
        paren_depth = ctx.paren_depth
        has_decimal = ctx.has_decimal

        if tok == "(":
            paren_depth += 1
            has_decimal  = False
        elif tok == ")":
            paren_depth -= 1
        elif token_type in {"Exp", "Term", "Power"}:
            has_decimal = False
        elif tok == ".":
            has_decimal = True

        ctx = ParseContext(trailing=tok, paren_depth=paren_depth,
                           has_decimal=has_decimal)

    return results


# ── Valid sequences ───────────────────────────────────────────
class TestValidSequences:

    def test_simple_addition(self):
        results = feed(build_chain(), ["5", "+", "3", "="])
        assert [r[0] for r in results] == ["Digit", "Exp", "Digit", "Solve"]

    def test_order_of_operations(self):
        results = feed(build_chain(), ["2", "+", "3", "*", "4", "="])
        assert [r[0] for r in results] == ["Digit", "Exp", "Digit", "Term", "Digit", "Solve"]

    def test_parenthesised_expression(self):
        results = feed(build_chain(), ["(", "6", "-", "2", ")", "*", "3", "="])
        assert [r[0] for r in results] == [
            "Factor", "Digit", "Exp", "Digit", "Factor", "Term", "Digit", "Solve"
        ]

    def test_power_expression(self):
        results = feed(build_chain(), ["2", "**", "8", "="])
        assert [r[0] for r in results] == ["Digit", "Power", "Digit", "Solve"]

    def test_function_wrapping_digit(self):
        results = feed(build_chain(), ["abs", "9", "="])
        assert [r[0] for r in results] == ["Function", "Digit", "Solve"]

    def test_nested_functions(self):
        results = feed(build_chain(), ["neg", "abs", "7", "="])
        assert [r[0] for r in results] == ["Function", "Function", "Digit", "Solve"]

    def test_decimal_number(self):
        results = feed(build_chain(), ["3", ".", "1", "4", "="])
        assert [r[0] for r in results] == ["Digit", "SoloDec", "Digit", "Digit", "Solve"]

    def test_floor_divide(self):
        results = feed(build_chain(), ["1", "7", "//", "5", "="])
        assert [r[0] for r in results] == ["Digit", "Digit", "Term", "Digit", "Solve"]

    def test_modulo(self):
        results = feed(build_chain(), ["1", "7", "%", "5", "="])
        assert [r[0] for r in results] == ["Digit", "Digit", "Term", "Digit", "Solve"]

    def test_complex_equation(self):
        # 234 + (12 / 3)
        results = feed(build_chain(),
                       ["2", "3", "4", "+", "(", "1", "2", "/", "3", ")", "="])
        assert [r[0] for r in results] == [
            "Digit", "Digit", "Digit",
            "Exp",
            "Factor",
            "Digit", "Digit",
            "Term",
            "Digit",
            "Factor",
            "Solve",
        ]

    def test_solve_only(self):
        assert build_chain().handle("=", ParseContext()) == ("Solve", "=")


# ── Invalid sequences ─────────────────────────────────────────
class TestInvalidSequences:

    def test_leading_plus_rejected(self):
        assert feed(build_chain(), ["+", "5"])[0] is None

    def test_double_operator_rejected(self):
        results = feed(build_chain(), ["1", "2", "+", "+"])
        assert results[3] is None
        assert results[0] is not None

    def test_operator_after_open_paren_rejected(self):
        assert feed(build_chain(), ["(", "+"])[1] is None

    def test_unmatched_close_paren_rejected(self):
        assert build_chain().handle(")", ParseContext(paren_depth=0)) is None

    def test_second_decimal_rejected(self):
        assert feed(build_chain(), ["1", ".", "5", "."])[3] is None

    def test_digit_after_close_paren_rejected(self):
        assert feed(build_chain(), ["(", "3", ")", "5"])[3] is None

    def test_unknown_token_rejected(self):
        assert build_chain().handle("xyz", ParseContext(trailing="5")) is None

    def test_leading_multiply_rejected(self):
        assert build_chain().handle("*", ParseContext()) is None


# ── Processed token values ────────────────────────────────────
class TestProcessedTokenValues:

    def test_exp_adds_spaces(self):
        _, processed = build_chain().handle("+", ParseContext(trailing="5"))
        assert processed == " + "

    def test_term_adds_spaces(self):
        _, processed = build_chain().handle("*", ParseContext(trailing="3"))
        assert processed == " * "

    def test_power_adds_spaces(self):
        _, processed = build_chain().handle("**", ParseContext(trailing="2"))
        assert processed == " ** "

    def test_factor_open_no_spaces(self):
        _, processed = build_chain().handle("(", ParseContext())
        assert processed == "("

    def test_solo_decimal_prefixes_zero(self):
        _, processed = build_chain().handle(".", ParseContext())
        assert processed == "0."

    def test_digit_returned_as_is(self):
        _, processed = build_chain().handle("7", ParseContext())
        assert processed == "7"


# ── Chain extension ───────────────────────────────────────────
class TestChainExtension:

    def test_new_function_accepted_in_extended_chain(self):
        extended = build_chain(functions=DEFAULT_FUNCTIONS | {"cbrt"})
        assert extended.handle("cbrt", ParseContext(trailing="8")) == ("Function", "cbrt")

    def test_new_function_rejected_in_default_chain(self):
        assert build_chain().handle("cbrt", ParseContext(trailing="8")) is None

    def test_existing_functions_still_work_after_extension(self):
        extended = build_chain(functions=DEFAULT_FUNCTIONS | {"cbrt"})
        for fn in DEFAULT_FUNCTIONS:
            assert extended.handle(fn, ParseContext(trailing="5")) == ("Function", fn)


# ── Pipeline: CoR → TreeBuilder → Interpreter ────────────────
class TestPipeline:
    """
    Full end-to-end: token list → parse_and_build() → .interpret()
    Validates that CoR validation, tree construction, and Interpreter
    evaluation all work together correctly.
    """

    ctx = _interp.Context()   # empty context (no variables needed)

    def _eval(self, tokens):
        return parse_and_build(tokens).interpret(self.ctx)

    def test_simple_addition(self):
        assert self._eval(["3", "+", "4", "="]) == 7

    def test_simple_subtraction(self):
        assert self._eval(["9", "-", "4", "="]) == 5

    def test_simple_multiplication(self):
        assert self._eval(["6", "*", "7", "="]) == 42

    def test_simple_division(self):
        assert self._eval(["8", "/", "4", "="]) == 2

    def test_operator_precedence_mul_over_add(self):
        # 2 + 3 * 4  →  2 + 12  =  14  (not (2+3)*4=20)
        assert self._eval(["2", "+", "3", "*", "4", "="]) == 14

    def test_parentheses_override_precedence(self):
        # (2 + 3) * 4  =  20
        assert self._eval(["(", "2", "+", "3", ")", "*", "4", "="]) == 20

    def test_power(self):
        assert self._eval(["2", "**", "8", "="]) == 256

    def test_floor_divide(self):
        assert self._eval(["1", "7", "//", "5", "="]) == 3

    def test_modulo(self):
        assert self._eval(["1", "7", "%", "5", "="]) == 2

    def test_unary_sqrt(self):
        assert self._eval(["sqrt", "9", "="]) == 3.0

    def test_unary_abs_of_negative(self):
        assert self._eval(["abs", "neg", "5", "="]) == 5.0

    def test_nested_unary_functions(self):
        # abs(neg(abs(neg(7)))) = 7
        assert self._eval(["abs", "neg", "abs", "neg", "7", "="]) == 7.0

    def test_function_over_group(self):
        # sqrt((3**2 + 4**2))  →  sqrt(9 + 16) = 5  — Pythagorean triple
        tokens = ["sqrt", "(", "3", "**", "2", "+", "4", "**", "2", ")", "="]
        assert self._eval(tokens) == 5.0

    def test_multidigit_number(self):
        # 234 + 0  =  234
        assert self._eval(["2", "3", "4", "+", "0", "="]) == 234

    def test_decimal_number(self):
        result = self._eval(["3", ".", "1", "4", "+", "0", "="])
        assert abs(result - 3.14) < 1e-9

    def test_invalid_token_raises(self):
        with pytest.raises(ValueError, match="Invalid token"):
            parse_and_build(["+", "5"])

    def test_tree_str_reflects_structure(self):
        # (2 + 3) * 4  →  tree str should show multiplication at root
        tree = parse_and_build(["(", "2", "+", "3", ")", "*", "4", "="])
        s = str(tree)
        assert "*" in s
        assert "+" in s
